drivers/sensor/adi/adxl345/adxl345_decoder.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2024 Analog Devices Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "adxl345.h"
+
+#ifdef CONFIG_ADXL345_STREAM
+
+#define SENSOR_SCALING_FACTOR (SENSOR_G / (16 * 1000 / 100))
+
+static const uint32_t accel_period_ns[] = {
+	[ADXL345_ODR_12HZ] = UINT32_C(1000000000) / 12,
+	[ADXL345_ODR_25HZ] = UINT32_C(1000000000) / 25,
+	[ADXL345_ODR_50HZ] = UINT32_C(1000000000) / 50,
+	[ADXL345_ODR_100HZ] = UINT32_C(1000000000) / 100,
+	[ADXL345_ODR_200HZ] = UINT32_C(1000000000) / 200,
+	[ADXL345_ODR_400HZ] = UINT32_C(1000000000) / 400,
+};
+
+static inline void adxl345_accel_convert_q31(q31_t *out, uint16_t sample)
+{
+	if (sample & BIT(9)) {
+		sample |= ADXL345_COMPLEMENT;
+	}
+	int32_t micro_ms2 = ((sample * SENSOR_G) / 32);
+	*out = CLAMP((((int64_t)micro_ms2) + (micro_ms2 % 1000000)), INT32_MIN, INT32_MAX);
```

Comment:
Similar to https://github.com/zephyrproject-rtos/zephyr/pull/78941#discussion_r1811265047